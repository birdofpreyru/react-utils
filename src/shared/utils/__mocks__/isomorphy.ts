/**
 * Mockup of the isomorphy module.
 */
/* global jest */

import * as IsomorphyM from '../isomorphy';

const MOCK_BUILD_TIMESTAMP = 'Wed, 29 Nov 2017 07:40:00 GMT';

const mock = jest.requireActual<typeof IsomorphyM>('../isomorphy');

mock.buildTimestamp = () => MOCK_BUILD_TIMESTAMP;

module.exports = mock;
