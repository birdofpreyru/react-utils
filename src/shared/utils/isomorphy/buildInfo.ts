// Encapsulates access to "Build Info" data.

// BEWARE: This should match the type of build info object generated by
// Webpack build (see "/config/webpack/app-base.js"), and currently this
// match is not checked automatically.
export type BuildInfoT = {
  key: string;
  publicPath: string;
  timestamp: string;
  useServiceWorker: boolean;
};

// Depending on the build mode & environment, BUILD_INFO is either a global
// variable defined at the app launch, or it is replaced by the actual value
// by the Webpack build.
declare const BUILD_INFO: BuildInfoT | undefined;

let buildInfo: BuildInfoT | undefined;

// On the client side "BUILD_INFO" should be injected by Webpack. Note, however,
// that in test environment we may need situations were environment is mocked as
// client-side, although no proper Webpack compilation is executed, thus no info
// injected; because of this we don't do a hard environment check here.
if (typeof BUILD_INFO !== 'undefined') buildInfo = BUILD_INFO;

/**
 * In scenarious where "BUILD_INFO" is not injected by Webpack (server-side,
 * tests, etc.) we expect the host codebase to explicitly set it before it is
 * ever requested. As a precaution, this function throws if build info has been
 * set already, unless `force` flag is explicitly set.
 * @param info
 * @param force
 */
export function setBuildInfo(info?: BuildInfoT, force = false): void {
  if (buildInfo !== undefined && !force) {
    throw Error('"Build Info" is already initialized');
  }
  buildInfo = info;
}

/**
 * Returns "Build Info" object; throws if it has not been initialized yet.
 * @returns
 */
export function getBuildInfo(): BuildInfoT {
  if (buildInfo === undefined) {
    throw Error('"Build Info" has not been initialized yet');
  }
  return buildInfo;
}
