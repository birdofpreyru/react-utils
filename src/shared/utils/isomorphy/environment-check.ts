// Checks for client- vs. server-side environment detection.

/**
 * `true` within client-side environment (browser), `false` at server-side.
 */
export const IS_CLIENT_SIDE: boolean = typeof process !== 'object'
  || !process.versions?.node
  || !!global.REACT_UTILS_FORCE_CLIENT_SIDE;

/**
 * `true` within the server-side environment (node), `false` at client-side.
 */
export const IS_SERVER_SIDE: boolean = !IS_CLIENT_SIDE;
