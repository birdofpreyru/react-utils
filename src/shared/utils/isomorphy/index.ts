import { getBuildInfo } from './buildInfo';
import { IS_CLIENT_SIDE, IS_SERVER_SIDE } from './environment-check';

/**
 * @ignore
 * @return {string} Code mode: "development" or "production".
 */
function getMode() {
  return process.env.NODE_ENV;
}

/**
 * Returns `true` if development version of the code is running;
 * `false` otherwise.
 */
export function isDevBuild(): boolean {
  return getMode() === 'development';
}

/**
 * Returns `true` if production build of the code is running;
 * `false` otherwise.
 */
export function isProdBuild(): boolean {
  return getMode() === 'production';
}

/**
 * Returns build timestamp of the front-end JS bundle.
 * @return ISO date/time string.
 */
export function buildTimestamp(): string {
  return getBuildInfo().timestamp;
}

export { IS_CLIENT_SIDE, IS_SERVER_SIDE, getBuildInfo };
