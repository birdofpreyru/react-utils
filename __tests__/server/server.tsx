import type { NextFunction, Request, Response } from 'express';

import type ServeFaviconM from 'serve-favicon';
import supertest from 'supertest';
import type WebpackM from 'webpack';
import type WebpackHotM from 'webpack-hot-middleware';

import serverFactory, { type CspOptionsT } from 'server/server';
import { setBuildInfo } from 'utils/isomorphy/buildInfo';

function noop() {
  // NOOP
}

// Test logger, which omits regular info message from the console.
const logger = {
  debug: noop,
  error: noop,
  info: noop,
  log: noop,
  warn: noop,
};

jest.mock<typeof ServeFaviconM>('serve-favicon', () => jest.fn(
  () => (req: Request, res: Response, next: NextFunction) => {
    next();
  },
));

jest.mock<typeof WebpackM>('webpack', () => {
  const mock = () => ({
    apply: jest.fn(),
    plugin: jest.fn(),
    watch: jest.fn(),
  });
  mock.ProgressPlugin = Object;
  return mock as unknown as typeof WebpackM;
});

jest.mock<typeof WebpackHotM>(
  'webpack-hot-middleware',
  () => jest.fn(
    () => (req: Request, res: Response, next: NextFunction) => {
      next();
    },
  ) as unknown as typeof WebpackHotM,
);

const TEST_CONTEXT = `${__dirname}/test_data`;

const TEST_FAVICON_PATH = '/path/to/favicon.ico';

const TEST_WEBPACK_CONFIG = {
  context: TEST_CONTEXT,
  output: {
    path: '/path/to/assets',
    publicPath: '/test/public/path/',
  },
};

afterEach(() => {
  delete process.env.DEV_TOOLS;
  setBuildInfo(undefined, true);
  jest.resetAllMocks();
});

test('Favicon support', async () => {
  const server = serverFactory(TEST_WEBPACK_CONFIG, {
    favicon: TEST_FAVICON_PATH,
    logger,
  });
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  expect(require('serve-favicon')).toHaveBeenCalledWith(TEST_FAVICON_PATH);
  await server;
});

test('Launch with dev tools', async () => {
  process.env.DEV_TOOLS = '1';
  const server = serverFactory(TEST_WEBPACK_CONFIG, { logger });
  await server;
});

describe('Server is functional', () => {
  let server: supertest.Agent;

  beforeAll(async () => {
    server = supertest(
      await serverFactory(TEST_WEBPACK_CONFIG, {
        Application: () => <div>Hello World!</div>,
        cspSettingsHook: (csp: CspOptionsT) => {
          const sources = csp.directives?.['default-src'] as string[];
          sources.push('https://sample.url');
          return csp;
        },
        logger,
        maxSsrRounds: 3,

        // Use cache in tests as well.
        staticCacheController: () => ({ key: 'key' }),
      }),
    );
  });

  test('Simple request', async () => {
    await server.get('/').expect(200)
      .expect((res) => {
        expect(res.headers['content-security-policy']).toMatchSnapshot();
        expect(res.text).toMatchSnapshot();
      });
  });
});
