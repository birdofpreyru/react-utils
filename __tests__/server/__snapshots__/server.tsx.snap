// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Server is functional Simple request 1`] = `"default-src 'self' https://sample.url;base-uri 'self';font-src 'self' data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self' 'unsafe-eval' 'nonce-00000000-0000-0000-0000-000000000001';script-src-attr 'none';style-src 'self' 'unsafe-inline';frame-src 'self' https://*.youtube.com;upgrade-insecure-requests"`;

exports[`Server is functional Simple request 2`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            
            <title data-rh="true"></title>
            
            <meta name="theme-color" content="#FFFFFF">
            
            
            
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghgc5WRH69eBKHiBZQmMDPrDIZjS0mo8ysGCjayxGzD6WmL8eLQdJGGbkbV9YPqFyABasN0sKSQcvlSF0QOvH+XQ==">
          </head>
          <body>
            
            <div id="react-view"><div>Hello World!</div></div>
            
            
          </body>
        </html>"
`;
